### ISSUE TITLE
Ukvm can't load unikernels with non-page-aligned ELF segments

### ISSUE THREAD

**Issue by ricarkol** [hioa-cs/IncludeOS#1435]
When loading the unikernel ELF, ukvm changes protection for the unikernel's memory pages. ukvm does this for each ELF segment; for example, executable segments are read+execute only. Now, the issue is that Unikernels with non memory aligned segments like the following (from an includeos unikernel) result in a failed `mprotect` and can't be loaded:

```
Program Headers:
  Type           Offset             VirtAddr           PhysAddr
                 FileSiz            MemSiz              Flags  Align
  LOAD           0x0000000000000120 0x0000000000a00120 0x0000000000a00000
                 0x000000000029d344 0x000000000029d344  RWE    20
  LOAD           0x000000000029d480 0x0000000001c2fb80 0x0000000001c2fb80 <=== THIS ONE
                 0x0000000000000000 0x000000000000a658  RW     80
  GNU_EH_FRAME   0x00000000001e9fb0 0x0000000000be9fb0 0x0000000000be9fb0
                 0x0000000000007edc 0x0000000000007edc  R      4
  GNU_STACK      0x0000000000000000 0x0000000000000000 0x0000000000000000
                 0x0000000000000000 0x0000000000000000  RWE    10

 Section to Segment mapping:
  Segment Sections...
   00     .multiboot .text .init .fini .got .ctors .dtors .init_array .rodata .eh_frame_hdr .eh_frame .gcc_except_table .data .jcr .tdata .elf_symbols 
   01     .bss <== this is not aligned
   02     .eh_frame_hdr 
   03  
```
`mprotect` of segment `01` (at `0x0000000001c2fb80`) fails because it's not page aligned.

This can be fixed on the IncludeOS or the solo5 side, and it's discussed further at: https://github.com/Solo5/solo5/issues/212

**Comment by fwsGonzo** [comment]
We want to enable protection via paging too, but not for individual sections. Instead we will protect the read-only data area as a whole, and it should be page-aligned. The data sections should also, as a whole, be aligned.

**Comment by ricarkol** [comment]
@fwsGonzo the protection you proposed would be based on page tables on the guest side right? Just in case, the one from ukvm is not using guest page tables, it's `mprotect`'ing the allocated memory that's given to kvm. So read only pages will be read only even if the guest page tables say they are writable.

**Comment by fwsGonzo** [comment]
> the protection you proposed would be based on page tables on the guest side right?
Yea, but only on PC, not on ukvm guest

**Comment by fwsGonzo** [comment]
I've looked at the BSS 16MB "global variable" area where solo5 puts things. It can work much better with IncludeOS if solo5 puts the global variable area at the end of BSS.

The beginning of the BSS is already used at start by IncludeOS as the symbol table area. IncludeOS then moves the symbols to the end of the BSS during start. The heap is then placed somewhere after symbols. In the solo5 kernel_start we can just add 16MB to the distance between <BSS, ELF symbols> and <heap>.

**Comment by fwsGonzo** [comment]
Should be solved with this https://github.com/hioa-cs/IncludeOS/pull/1462/commits/8c69784deb41446183f3f420795aac00d6633c27

**Comment by alfred-bratterud** [comment]
@ricarkol I've now made chainloader and legacy bootloader ignore the ELF header and load the binary starting at the first loadable sebment ( #1472 ). Could you have a look and see if this fixes things?

**Comment by ricarkol** [comment]
Hi @alfred-bratterud , I'm not able to build solo5 unikernels on the latest `dev`, getting this:
```
[100%] Built target service
Scanning dependencies of target prepend_bootloader
vmbuild: /root/IncludeOS/vmbuild/vmbuild.cpp:261: int main(int, char **): Assertion `multiboot_hdr->entry_addr == srv_entry' failed.
```
Does `./test_ukvm.sh` work on your side?

**Comment by alfred-bratterud** [comment]
Hi @ricarkol, the assert in vmbuild wants to verify that the multiboot header for the ELF binary is correct. It seems the multiboot header in solo5's `start.asm` was missing the entry address, whch is why the assert failed. 

Vmbuild produces a bootable binary with our legacy bootloader and isn't really necessary for ukvm - and neither is multiboot as you get all the information you need directly from the ELF. However since you already had a partial multiboot entry I thought the easiest fix was to just add the entry point. #1477 fixed the issue for me and I was able to boot the example service with `test_ukvm.sh`.

Great if you could confirm that it works for you now - and ideally that we can now use upstream Solo5.

**Comment by ricarkol** [comment]
Awesome, thanks @alfred-bratterud . It builds correctly now, and the first segment is aligned `0x0000000000a00000`. Will get a PR pointing to the latest solo5 master (after testing and all).

**Comment by alfred-bratterud** [comment]
Great @ricarkol , let me know if there's anything else we can do to help!

**Comment by alfred-bratterud** [comment]
Looks like this issue is now solved, see #1478
